lessons/basics/control-structures.md

```diff
@@ -128,3 +128,49 @@ iex> cond do
 ...> end
 "Catch all"
 ```
+
+## `with`
+
+The special form `with` is useful when you might use a nested `case` statement or situations that cannot cleanly be piped together. We'll start with a simple example of `with` and then look at something more significant:
+
+```elixir
+iex> user = %{first: "Sean", last: "Callan"}
+iex> with {:ok, first} <- Map.fetch(user, :first),
+...>      {:ok, last} <- Map.fetch(user, :last),
+...>      do: last <> ", " <> first
+"Callan, Sean"
+```
+
+In the event that an expression fails to match, the non-matching value will be returned:
+
+```elixir
+iex> user = %{first: "doomspork"}
+%{first: "doomspork"}
+iex> with {:ok, first} <- Map.fetch(user, :first),
+...>      {:ok, last} <- Map.fetch(user, :last),
+...>      do: last <> ", " <> first
+:error
```

Comment:
I see we get an error here but would be good showing also an "real alternative without errors"